syllabus/core/stat_recorder.py

```diff
@@ -34,45 +31,96 @@ def record(self, episode_return: float, episode_length: int, episode_task, env_i
         """
         Records the length and return of an episode for a given task.
 
-        :param task: Identifier for the task
+        :param episode_task: Identifier for the task
         :param episode_length: Length of the episode, i.e. the total number of steps taken during the episode
         :param episodic_return: Total return for the episode
         """
 
         if episode_task in self.tasks:
-            self.records[episode_task].append({
-                "r": episode_return,
-                "l": episode_length,
-                "env_id": env_id
-            })
-            self.stats[episode_task]['mean_r'] = np.mean([record["r"] for record in self.records[episode_task]])
-            self.stats[episode_task]['var_r'] = np.var([record["r"] for record in self.records[episode_task]])
-            self.stats[episode_task]['mean_l'] = np.mean([record["l"] for record in self.records[episode_task]])
-            self.stats[episode_task]['var_l'] = np.var([record["l"] for record in self.records[episode_task]])
+            if self.calc_past_N:
+                self.records[episode_task].append({
+                    "r": episode_return,
+                    "l": episode_length,
+                    "env_id": env_id
+                })
+                self.records[episode_task] = self.records[episode_task][-self.calc_past_N:]
+                self.stats[episode_task]['mean_r'] = np.mean([record["r"] for record in self.records[episode_task]])
+                self.stats[episode_task]['var_r'] = np.var([record["r"] for record in self.records[episode_task]])
+                self.stats[episode_task]['mean_l'] = np.mean([record["l"] for record in self.records[episode_task]])
+                self.stats[episode_task]['var_l'] = np.var([record["l"] for record in self.records[episode_task]])
+                # only save mean/variance the past N episodes
+            else:
+                # save the mean/variance of all the episodes
+                if 'mean_r' not in self.stats[episode_task].keys():
+                    # the first episode for a task
+                    self.stats[episode_task]['mean_r'] = episode_return
+                    self.stats[episode_task]['mean_r_squared'] = episode_return ** 2
+                    self.stats[episode_task]['var_r'] = 0
+                    self.stats[episode_task]['mean_l'] = episode_length
+                    self.stats[episode_task]['mean_l_squared'] = episode_length ** 2
+                    self.stats[episode_task]['var_l'] = 0
+                else:
+                    N_past = len(self.records[episode_task])
+                    
+                    self.stats[episode_task]['mean_r'] =round((self.stats[episode_task]['mean_r'] * N_past + episode_return) / (N_past + 1), 4)
```

Comment:
We probably shouldn't round the saved values, we should only round them when logging or printing